PR: https://github.com/odoo/odoo/pull/236731

From: a063ce22258623beb07ad7f3323c053108d0c485
From: Léopold Cantraine
Date: 2025-12-19 16:15:34

Breaking data model changes scores: del:7 + add:4, change matches:
-    printer_ids = fields.Many2many('pos.printer', 'pos_config_printer_rel', 'config_id', 'printer_id', string='Order Printers')
+    preparation_printer_ids = fields.Many2many('pos.printer', 'pos_config_printer_rel', 'config_id', 'printer_id', string="Preparation Printers", domain="[('use_type', '=', 'preparation')]")
+    receipt_printer_ids = fields.Many2many('pos.printer', 'pos_config_receipt_printer_rel', 'config_id', 'printer_id', string="Receipt Printers", domain="[('use_type', '=', 'receipt')]")
-    epson_printer_ip = fields.Char(
-    pos_use_order_printer = fields.Boolean(compute='_compute_pos_printer', store=True, readonly=False)
-    pos_printer_ids = fields.Many2many(related='pos_config_id.printer_ids', readonly=False)
+    pos_receipt_printer_ids = fields.Many2many(related='pos_config_id.receipt_printer_ids', domain="[('use_type', '=', 'receipt')]", readonly=False)
+    pos_preparation_printer_ids = fields.Many2many(related='pos_config_id.preparation_printer_ids', domain="[('use_type', '=', 'preparation')]", readonly=False)
-    pos_epson_printer_ip = fields.Char(related='pos_config_id.epson_printer_ip', readonly=False)

Total Changes: 58

[IMP] pos_*, point_of_sale, iot_drivers: Unify printer handling logic across POS components

*: pos_enterprise, pos_imin, pos_iot, pos_restaurant

The goal of this commit was to unify the printer logic accross the POS. Before this commit, preparation and receipt was handled differently. Preparation printers were handled in the `pos.printer` model in a dedicated view.
And receipt printers were handled in the `pos.config` model in the pos.config settings view. Here, printers connected through the iot box and the EPos ones were handled under two different checkbox settings. It was a bit difficult to handle them if a client have a config with preparation printers and receipt printers connected through maneers.

The goal was also two permit a client to have multiple receipt printers for one POS. Before, only one was permitted. And a printer can also be used by multiple POS config.

After this commit,

* `pos.printer` model is used by every single printer in the POS. The configuration page is located at one place and the pos config setting look a better simple. We can configure one or multiple receipt printers for one pos config in the settings and a default one. The default one is used by default by the POS to print receipt. If the printer isn't accessible (lack of paper, loss of connection,...), a modal appears to offer the client to retry, download, continue (as before this commit) with a last option to choose a fallback printer if one is available in the pos config.

* two test buttons were created to test a specified printer in the `pos.printer`view and every printer available for a pos config in the `pos.config` settings view (iot and epson ones). On the ticket printed, we can identify easily the printer with its name.

task : 5266055

closes odoo/odoo#236731

Related: odoo/enterprise#100059
Related: odoo/upgrade#9053
Signed-off-by: Adrien Guilliams (adgu) <adgu@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_config.py
+++ b/addons/point_of_sale/models/pos_config.py
@@ -69,7 +69,9 @@ class PosConfig(models.Model):
         return tip_product_id
 
     name = fields.Char(string='Point of Sale', required=True, help="An internal identification of the point of sale.")
-    printer_ids = fields.Many2many('pos.printer', 'pos_config_printer_rel', 'config_id', 'printer_id', string='Order Printers')
+    preparation_printer_ids = fields.Many2many('pos.printer', 'pos_config_printer_rel', 'config_id', 'printer_id', string="Preparation Printers", domain="[('use_type', '=', 'preparation')]")
+    receipt_printer_ids = fields.Many2many('pos.printer', 'pos_config_receipt_printer_rel', 'config_id', 'printer_id', string="Receipt Printers", domain="[('use_type', '=', 'receipt')]")
+    default_receipt_printer_id = fields.Many2one('pos.printer', string="Default Receipt Printer")
     use_order_printer = fields.Boolean('Order Printer')
     is_installed_account_accountant = fields.Boolean(string="Is the Full Accounting Installed",
         compute="_compute_is_installed_account_accountant")
@@ -166,6 +168,7 @@ class PosConfig(models.Model):
     company_has_template = fields.Boolean(string="Company has chart of accounts", compute="_compute_company_has_template")
     current_user_id = fields.Many2one('res.users', string='Current Session Responsible', compute='_compute_current_session_user')
     other_devices = fields.Boolean(string="Other Devices", help="Connect devices to your PoS without an IoT Box.")
+    preparation_devices = fields.Boolean(string="Preparation devices", help="Connect preparation printers to print to the bar, kitchen,...")
     rounding_method = fields.Many2one('account.cash.rounding', string="Cash rounding")
     cash_rounding = fields.Boolean(string="Cash Rounding")
     only_round_cash_method = fields.Boolean(string="Only apply rounding on cash")
@@ -195,13 +198,6 @@ class PosConfig(models.Model):
     order_edit_tracking = fields.Boolean(string="Track orders edits", help="Store edited orders in the backend", default=False)
     last_data_change = fields.Datetime(string='Last Write Date', readonly=True, compute='_compute_local_data_integrity', store=True)
     fallback_nomenclature_id = fields.Many2one('barcode.nomenclature', string="Fallback Nomenclature")
-    epson_printer_ip = fields.Char(
-        string='Epson Printer IP',
-        help=(
-            "Local IP address of an Epson receipt printer, or its serial number if the "
-            "'Automatic Certificate Update' option is enabled in the printer settings."
-        ),
-    )
     use_fast_payment = fields.Boolean('Fast Payment Validation', help="Enable fast payment methods to validate orders on the product screen.")
     fast_payment_method_ids = fields.Many2many(
         'pos.payment.method', string='Fast Payment Methods', compute="_compute_fast_payment_method_ids", relation='pos_payment_method_config_fast_validation_relation',
@@ -624,7 +620,7 @@ class PosConfig(models.Model):
         self._check_header_footer(vals)
         self._reset_default_on_vals(vals)
         if ('use_order_printer' in vals and not vals['use_order_printer']):
-            vals['printer_ids'] = [fields.Command.clear()]
+            vals['preparation_printer_ids'] = [fields.Command.clear()]
 
         bypass_payment_method_ids_forbidden_change = self.env.context.get('bypass_payment_method_ids_forbidden_change', False)
 
@@ -1214,9 +1210,3 @@ class PosConfig(models.Model):
 
     def _is_quantities_set(self):
         return self.use_closing_entry_by_product
-
-    @api.onchange("epson_printer_ip")
-    def _onchange_epson_printer_ip(self):
-        for rec in self:
-            if rec.epson_printer_ip:
-                rec.epson_printer_ip = format_epson_certified_domain(rec.epson_printer_ip)

--- a/addons/point_of_sale/models/pos_printer.py
+++ b/addons/point_of_sale/models/pos_printer.py
@@ -36,27 +36,31 @@ class PosPrinter(models.Model):
         string='Printer Type',
         default='epson_epos',
         selection=[
-            ('epson_epos', 'Use an Epson printer'),
+            ('epson_epos', 'IP address'),
         ]
     )
+    use_type = fields.Selection(selection=[
+        ('preparation', "Preparation"),
+        ('receipt', "Receipt"),
+    ], string="Type", default="preparation")
     product_categories_ids = fields.Many2many('pos.category', 'printer_category_rel', 'printer_id', 'category_id', string='Printed Product Categories')
-    pos_config_ids = fields.Many2many('pos.config', 'pos_config_printer_rel', 'printer_id', 'config_id')
+    pos_config_ids = fields.Many2many('pos.config', 'pos_config_receipt_printer_rel', 'printer_id', 'config_id', string="Point of Sale")
     epson_printer_ip = fields.Char(
         string='Epson Printer IP Address',
         help=(
             "Local IP address of an Epson receipt printer, or its serial number if the "
             "'Automatic Certificate Update' option is enabled in the printer settings."
         ),
-        default="0.0.0.0"
     )
+    use_lna = fields.Boolean(string="Use Local Network Access")
 
     @api.model
     def _load_pos_data_domain(self, data, config):
-        return [('id', 'in', config.printer_ids.ids)]
+        return [('id', 'in', config.preparation_printer_ids.ids + config.receipt_printer_ids.ids)]
 
     @api.model
     def _load_pos_data_fields(self, config):
-        return ['id', 'name', 'product_categories_ids', 'printer_type', 'epson_printer_ip']
+        return ['id', 'name', 'product_categories_ids', 'printer_type', 'use_type', 'epson_printer_ip']
 
     @api.constrains('epson_printer_ip')
     def _constrains_epson_printer_ip(self):

--- a/addons/point_of_sale/models/res_config_settings.py
+++ b/addons/point_of_sale/models/res_config_settings.py
@@ -54,8 +54,9 @@ class ResConfigSettings(models.TransientModel):
     pos_module_pos_restaurant = fields.Boolean(related='pos_config_id.module_pos_restaurant', readonly=False)
     pos_module_pos_appointment = fields.Boolean(related="pos_config_id.module_pos_appointment", readonly=False)
     pos_module_pos_avatax = fields.Boolean(related='pos_config_id.module_pos_avatax', readonly=False)
-    pos_use_order_printer = fields.Boolean(compute='_compute_pos_printer', store=True, readonly=False)
-    pos_printer_ids = fields.Many2many(related='pos_config_id.printer_ids', readonly=False)
+    pos_default_receipt_printer_id = fields.Many2one(related='pos_config_id.default_receipt_printer_id', readonly=False)
+    pos_receipt_printer_ids = fields.Many2many(related='pos_config_id.receipt_printer_ids', domain="[('use_type', '=', 'receipt')]", readonly=False)
+    pos_preparation_printer_ids = fields.Many2many(related='pos_config_id.preparation_printer_ids', domain="[('use_type', '=', 'preparation')]", readonly=False)
 
     pos_allowed_pricelist_ids = fields.Many2many('product.pricelist', compute='_compute_pos_allowed_pricelist_ids')
     pos_amount_authorized_diff = fields.Float(related='pos_config_id.amount_authorized_diff', readonly=False)
@@ -83,6 +84,7 @@ class ResConfigSettings(models.TransientModel):
     pos_manual_discount = fields.Boolean(related='pos_config_id.manual_discount', readonly=False)
     pos_only_round_cash_method = fields.Boolean(related='pos_config_id.only_round_cash_method', readonly=False)
     pos_other_devices = fields.Boolean(related='pos_config_id.other_devices', readonly=False)
+    pos_preparation_devices = fields.Boolean(related='pos_config_id.preparation_devices', readonly=False)
     pos_payment_method_ids = fields.Many2many(related='pos_config_id.payment_method_ids', readonly=False)
     pos_picking_policy = fields.Selection(related='pos_config_id.picking_policy', readonly=False)
     pos_picking_type_id = fields.Many2one(related='pos_config_id.picking_type_id', readonly=False)
@@ -114,7 +116,6 @@ class ResConfigSettings(models.TransientModel):
     pos_basic_receipt = fields.Boolean(related='pos_config_id.basic_receipt', readonly=False)
     pos_fallback_nomenclature_id = fields.Many2one(related='pos_config_id.fallback_nomenclature_id', domain="[('id', '!=', barcode_nomenclature_id)]", readonly=False)
     group_pos_preset = fields.Boolean(string="Presets", implied_group="point_of_sale.group_pos_preset", help="Hide or show the Presets menu in the Point of Sale configuration.")
-    pos_epson_printer_ip = fields.Char(related='pos_config_id.epson_printer_ip', readonly=False)
     pos_use_fast_payment = fields.Boolean(related='pos_config_id.use_fast_payment', readonly=False)
     pos_fast_payment_method_ids = fields.Many2many(related='pos_config_id.fast_payment_method_ids', readonly=False)
 
@@ -227,14 +228,7 @@ class ResConfigSettings(models.TransientModel):
 
     @api.model
     def _is_cashdrawer_displayed(self, res_config):
-        return res_config.pos_other_devices and bool(res_config.pos_epson_printer_ip)
-
-    @api.depends('pos_module_pos_restaurant', 'pos_config_id')
-    def _compute_pos_printer(self):
-        for res_config in self:
-            res_config.update({
-                'pos_use_order_printer': res_config.pos_config_id.use_order_printer,
-            })
+        return res_config.pos_other_devices and bool(res_config.pos_default_receipt_printer_id)
 
     @api.depends('pos_limit_categories', 'pos_config_id')
     def _compute_pos_iface_available_categ_ids(self):
@@ -252,7 +246,7 @@ class ResConfigSettings(models.TransientModel):
             else:
                 res_config.pos_selectable_categ_ids = self.env['pos.category'].search([])
 
-    @api.depends('pos_config_id', 'pos_epson_printer_ip', 'pos_other_devices')
+    @api.depends('pos_config_id', 'pos_default_receipt_printer_id', 'pos_other_devices')
     def _compute_pos_iface_cashdrawer(self):
         for res_config in self:
             if self._is_cashdrawer_displayed(res_config):
@@ -316,9 +310,3 @@ class ResConfigSettings(models.TransientModel):
                     old._add_trusted_config_id(config.pos_config_id)
                 if old.id in removed_trusted_configs:
                     old._remove_trusted_config_id(config.pos_config_id)
-
-    @api.onchange("pos_epson_printer_ip")
-    def _onchange_epson_printer_ip(self):
-        for rec in self:
-            if rec.pos_epson_printer_ip:
-                rec.pos_epson_printer_ip = format_epson_certified_domain(rec.pos_epson_printer_ip)
